In [ ]:
#!pip install missingno

In [123]:
import requests
import xmltodict
from datetime import datetime, timedelta
import os
import json
import sys

from sqlalchemy import create_engine
import pymysql
import pandas as pd

In [39]:
# DB 접속정보 설정
DB_URL = 'jdbc:mysql://localhost:3306/farmoney?useUnicode=true&characterEncoding=UTF-8'
DB_USER = 'hoit'
DB_PASS = '141008'

In [42]:
# pandas dataframe <-> mysql 용 connecntion
db_connection_str = 'mysql+pymysql://hoit:141008@localhost/farmoney'
db_connection = create_engine(db_connection_str)
conn = db_connection.connect()

## 최신데이터 수집
2021.09.17 ~

### 농업기상 조회기간별 일 기본 관측데이터 조회

In [75]:
certf = ['cRubis4JVspr1lTWwQzXYqKwYs4tRA7P5S1W7Vo%2Bi%2FxenmhOzGzusFs3XSInzRFLAW0j7IdxSIWywgQL3e9o5Q%3D%3D','Bai%2FvTgIbiRsY5w5O7aVL%2BLHplODFxh0r0aafEu2mjWuRQY0s70R3Kzl3gszDRRXDL128%2BYnn4wfcHeZ3jD0tQ%3D%3D','Cmgie9fneM4cfj2ViQV5T2SU9poQGtlmBhpVAztPgY4PTtArNthKa4VmCTKxNQNnWy4Kz3PKw6qwmRbfu2sXzw%3D%3D']

yesterday = datetime.today() - timedelta(1)
start = end = yesterday.strftime("%Y-%m-%d")
#start = end = '2021-10-04'

spot = ['신안군','진도군','영광군','완주군','부산시','청주시','아산시','구미시','평창군','남양주','용인시','양주시','김포시','김제시', '의성군','무안군','강진군','장성군','김해시','세종시','예산군','횡성군','광주시','이천시','고양시','안산시']

spot_code = ['535812A001','539823A001','513842A001','055365F016','618803A001','363844A001','336812A001','039102B003','232803A001','472830A001','449872A001','482841A001','415743A001','576040A001','037339B009','534833E001','059223A001','515804A001','621100A001','339814A001','340861A001','225874A001','464030A001','467800A001','412040A002','425320A001']

page = ['1','2','3','4']


for j in range(0, len(spot_code)):
    base_url = 'http://apis.data.go.kr/1390802/AgriWeather/WeatherObsrInfo/GnrlWeather/getWeatherTermDayList?'
    service = 'serviceKey={}'.format(certf[1])
    page_no = '&Page_No={}&Page_Size=100'.format(page[0]) 
    begin = '&begin_Date={}'.format(start)
    finish = '&end_Date={}'.format(end)
    obs = '&obsr_Spot_Code={}'.format(spot_code[j])

    url = base_url + service +page_no + begin + finish + obs
    response = requests.get(url)

    # ensure_ascii = False : 파일이 저장될때 한글이 유니코드 형태로 저장되지 않도록 함
    if response.status_code == 200:
        xd = xmltodict.parse(response.text) # XML 형식 데이터를 dict 형식으로 바꾼다
        xdj = json.dumps(xd, ensure_ascii=False) # dict형식의 데이터를 json 형식으로 변환
        xdjd = json.loads(xdj) # json형식의 데이터를 dict형식으로 변환

        file_path = "./data/일별농업기상/weather{}_{}.json".format(spot[j],start)
        with open(file_path, 'w', encoding='utf-8') as outfile:
            json.dump(xdjd, outfile, indent='\t', ensure_ascii=False)
        print(start,spot[j])
    else:
        print(response.status_code)

2021-10-05 신안군
2021-10-05 진도군
2021-10-05 영광군
2021-10-05 완주군
2021-10-05 부산시
2021-10-05 청주시
2021-10-05 아산시
2021-10-05 구미시
2021-10-05 평창군
2021-10-05 남양주
2021-10-05 용인시
2021-10-05 양주시
2021-10-05 김포시
2021-10-05 김제시
2021-10-05 의성군
2021-10-05 무안군
2021-10-05 강진군
2021-10-05 장성군
2021-10-05 김해시
2021-10-05 세종시
2021-10-05 예산군
2021-10-05 횡성군
2021-10-05 광주시
2021-10-05 이천시
2021-10-05 고양시
2021-10-05 안산시


In [80]:
# 데이터확인
file_path = './data/일별농업기상/weather{}_{}.json'.format('의성군','2021-10-04')
with open(file_path, "r", encoding='utf-8') as json_file:
    json_data = json.load(json_file)
    data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
    print(data)

{
	"response": {
		"header": {
			"result_Code": "200",
			"result_Msg": "OK"
		},
		"body": {
			"rcdcnt": "100",
			"page_No": "1",
			"total_Count": "1",
			"items": {
				"item": {
					"no": "1",
					"stn_Code": "037339B009",
					"stn_Name": "의성군 의성읍",
					"date": "2021-10-04",
					"temp": "19.7",
					"max_Temp": "28.6",
					"min_Temp": "12.4",
					"hum": "90.9",
					"widdir": "209.8",
					"wind": "0.4",
					"rain": "0",
					"sun_Time": "368",
					"sun_Qy": "12.9",
					"condens_Time": "923",
					"gr_Temp": null,
					"soil_Temp": "20.4",
					"soil_Wt": "50"
				}
			}
		}
	}
}


### 유통정보(시장도매인거래) - 품목별등급별가격

In [152]:
#start = datetime.strptime("20210915", "%Y%m%d")
#end = datetime.strptime("20210915", "%Y%m%d")

yesterday = datetime.today() - timedelta(1)
#yesterday = datetime.strptime('20211005', '%Y%m%d')
#start = end = yesterday.strftime("%Y-%m-%d")

# 당일
dates = yesterday.strftime("%Y%m%d")
#[(start + timedelta(days=x)).strftime("%Y%m%d") for x in range(0, 1+(end-start).days)]

# 하루전
dates_p = yesterday - timedelta(1)
dates_p = dates_p.strftime("%Y%m%d")
#[]
#for i in range(0, 1+(end-start).days):
#    pre = [(start + timedelta(days=x)) for x in range(0, 1+(end-start).days)][i] - timedelta(1)
#    dates_p.append(pre.strftime('%Y%m%d'))

# 일주전
dates_s = yesterday - timedelta(7)
dates_s = dates_s.strftime("%Y%m%d")
#dates_s = []
#for i in range(0, 1+(end-start).days):
#    sev = [(start + timedelta(days=x)) for x in range(0, 1+(end-start).days)][i] - timedelta(7)
#    dates_s.append(sev.strftime('%Y%m%d'))



base_url = 'http://www.garak.co.kr/publicdata/dataOpen.do?id=3043&passwd=multi1234!&dataid=data37&pagesize=10&pageidx=1&portal.templet=false'
date = '&s_date={}'.format(dates)
date_p = '&s_date_p={}'.format(dates_p)
date_s = '&s_date_p7={}'.format(dates_s)
product = '&s_pum_nm=2&s_pummok={}'.format('대파')

url = base_url + date + date_p + date_s + product
response = requests.get(url)

# ensure_ascii = False : 파일이 저장될때 한글이 유니코드 형태로 저장되지 않도록 함
if response.status_code == 200:
    xd = xmltodict.parse(response.text) # XML 형식 데이터를 dict 형식으로 바꾼다
    xdj = json.dumps(xd, ensure_ascii=False) # dict형식의 데이터를 json 형식으로 변환
    xdjd = json.loads(xdj) # json형식의 데이터를 dict형식으로 변환

    file_path = "./data/시장도매인가격/시장도매{}.json".format(dates)
    with open(file_path, 'w', encoding='utf-8') as outfile:
        json.dump(xdjd, outfile, indent='\t', ensure_ascii=False)
    print(dates)
else:
    print(response.status_code)

20211005


In [172]:
# 데이터확인
file_path = './data/시장도매인가격/시장도매{}.json'.format(dates)
with open(file_path, "r", encoding='utf-8') as json_file:
    json_data = json.load(json_file)
    data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
    print(file_path,'\n',data)

./data/시장도매인가격/시장도매20211005.json 
 {
	"lists": {
		"list_total_count": "4",
		"list": [
			{
				"PUM_NM": "대파(일반)",
				"G_NAME": "특",
				"U_NAME": "1 kg단",
				"MI_P": "2,050",
				"MA_P": "3,300",
				"AV_P": "2,711",
				"PAV_P": "1,955",
				"FLUC": "756",
				"D_MARK": "139.0",
				"J_7": "1,966",
				"J_7_MARK": "137.9"
			},
			{
				"PUM_NM": "대파(일반)",
				"G_NAME": "상",
				"U_NAME": "1 kg단",
				"MI_P": "1,300",
				"MA_P": "2,050",
				"AV_P": "1,450",
				"PAV_P": "1,361",
				"FLUC": "89",
				"D_MARK": "107.0",
				"J_7": "1,554",
				"J_7_MARK": "93.3"
			},
			{
				"PUM_NM": "대파(일반)",
				"G_NAME": "보통",
				"U_NAME": "1 kg단",
				"MI_P": "1,000",
				"MA_P": "1,300",
				"AV_P": "1,171",
				"PAV_P": "1,160",
				"FLUC": "11",
				"D_MARK": "101.0",
				"J_7": "1,223",
				"J_7_MARK": "95.8"
			},
			{
				"PUM_NM": "대파(일반)",
				"G_NAME": "하",
				"U_NAME": "1 kg단",
				"MI_P": "500",
				"MA_P": "1,000",
				"AV_P": "866",
				"PAV_P": "832",
				"FLUC": "34",
			

### 전국 도매시장 일별 정산 경락가격 상세정보

In [7]:
#!pip install tqdm

     |████████████████████████████████| 76 kB 1.7 MB/s eta 0:00:01


In [195]:
# import pymysql
import requests
# from bs4 import BeautifulSoup
from urllib3.util.retry import Retry
from requests.adapters import HTTPAdapter
import time
from tqdm.notebook import tqdm

s = requests.Session()
retries = Retry(total=5,
               backoff_factor=4, # 2, 4, 8, 16, 32
               status_forcelist=[500, 502, 503, 504])
headers= {
    'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:87.0) Gecko/20100101 Firefox/87.0'
}

def crawl(base_url, params, directory, name):
    today = datetime.datetime.now()
    res = s.get(base_url, headers=headers, params=params)
    if res.status_code == 200:
        result = json.loads(json.dumps(xmltodict.parse(res.content.decode('utf-8')), ensure_ascii=False))
        with open(directory + today.strftime("%Y%m%d%H%M%S") + name + ".json", "w", encoding='utf-8') as json_file:
            json.dump(result, json_file, ensure_ascii=False)
        return result
    else:
        return res.status_code
    
def crawl_json(base_url, params, directory, name):
    today = datetime.now()
    res = s.get(base_url, headers=headers, params=params)
    if res.status_code == 200:
        result = json.loads(res.content.decode('utf-8'))
        with open(directory + name + ".json", "w", encoding='utf-8') as json_file:
            json.dump(result, json_file, ensure_ascii=False)
        return result
    else:
        return res.status_code
keys = [
    '3316101bc570d80943c917dcaf175c59da913ddbb38519f9103d9dbc46d5aef7'
]

In [198]:
slash = [keys[0], 'json', 'Grid_20180118000000000581_1', 0, 0]

start_date_i = 0

#yesterday = datetime.today() - timedelta(1)
#start = end = datetime.strptime(yesterday.strftime("%Y%m%d"),"%Y%m%d")
#dates = [(start + timedelta(days=x)).strftime("%Y%m%d") for x in range(0, 1+(end-start).days)]

dates = ['20210930']

for date_i in tqdm(range(len(dates))):
    date = dates[date_i]
    if date_i < start_date_i:
        continue
    params = {
        'DELNG_DE': date,
        'PRDLST_CD': 1202,
    }
    time.sleep(0.3)
    slash[3] = str(1)
    slash[4] = str(1000)
    base_url = 'http://211.237.50.150:7080/openapi/' + '/'.join(slash)
    result = crawl_json(base_url, params, './data/도매시장경매가격/', '_'.join([date, slash[3], slash[4]]))
    if type(result) == int:
        print(date + ' error')
    end = result['Grid_20180118000000000581_1']['totalCnt'] // 1000 + 1
    for i in tqdm(range(1, end), leave=False):
        time.sleep(0.3)
        slash[3] = str(i*1000 + 1)
        slash[4] = str((i+1)*1000)
        base_url = 'http://211.237.50.150:7080/openapi/' + '/'.join(slash)
        crawl_json(base_url, params, './data/도매시장경매가격/', '_'.join([date, slash[3], slash[4]]))

  0%|          | 0/1 [00:00<?, ?it/s]

  0%|          | 0/1 [00:00<?, ?it/s]

In [199]:
# 데이터확인
file_path = './data/도매시장경매가격/{}_1_1000.json'.format(dates[0])
with open(file_path, "r", encoding='utf-8') as json_file:
    json_data = json.load(json_file)
    data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
    print(data)

{
	"Grid_20180118000000000581_1": {
		"totalCnt": 1413,
		"startRow": 1,
		"endRow": 1000,
		"result": {
			"code": "INFO-000",
			"message": "정상 처리되었습니다."
		},
		"row": [
			{
				"ROW_NUM": 1,
				"DELNG_DE": "20210930",
				"PBLMNG_WHSAL_MRKT_NM": "수원도매시장",
				"PBLMNG_WHSAL_MRKT_CD": "310101",
				"CPR_NM": "수원청과",
				"CPR_CD": "31010102",
				"RISENO": "01",
				"BIDTIME": "",
				"AUC_SE_CODE": "1",
				"AUC_SE_NM": "경매",
				"ORGNO": "0203003",
				"PRDLST_NM": "대파",
				"PRDLST_CD": "1202",
				"SPCIES_NM": "대파(일반)",
				"SPCIES_CD": "120201",
				"CPR_USE_PRDLST_NM": "파",
				"CPR_USE_SPCIES_NM": "대파",
				"PRICE": 550,
				"DELNGBUNDLE_QY": 1,
				"STNDRD": "kg 단 ",
				"STNDRD_CD": "121100",
				"GRAD": "없음",
				"GRAD_CD": "10",
				"SHIPMNT_SE_CODE": "5",
				"SHIPMNT_SE_NM": "수입",
				"SANJI_CD": "131600",
				"SANJI_NM": "서울 중랑구 서울중랑우체국사서함",
				"CPR_USE_SANJI_CD": "3236",
				"CPR_USE_SANJI_NM": "강원도 평창군",
				"DELNG_QY": 350
			},
			{
				"ROW_NUM": 2,
				"DELNG_DE"

## Spark로 dataframe 생성 -> MySQL에 저장

~현재까지 테이블로 저장

매일 새로 수집한 데이터는 기존 테이블 불러와서 아래에 붙인 후 db에 업데이트

새로 수집한 데이터(21.10.01~) 새 테이블로 저장

In [ ]:
# !pip install pyspark
# !pip install findspark

spark = SparkSession\
    .builder\
    .appName('sql')\
    .config('spark.driver.extraClassPath', '/home/opt/spark/jars/mysql-connector-java-8.0.26.jar')\
    .getOrCreate()

conf = SparkConf() \
    .setMaster('local') \
    .setAppName('sql') \
    .set('spark.driver.extraClassPath','jdbc/mysql-connector-java-8.0.26.jar')
sc = SparkContext(conf=conf)
spark = SparkSession(sc)

In [11]:
import re

In [12]:
import findspark
findspark.init()

import pyspark # only run after findspark.init()

from pyspark import SparkConf, SparkContext
from pyspark.sql import *
from pyspark.sql.functions import col, lit, get_json_object
from pyspark.sql.types import *

In [13]:
spark = SparkSession.builder.config('spark.driver.extraClassPath', 'jdbc/mysql-connector-java-8.0.26.jar').config('spark.executor.extraClassPath', 'jdbc/mysql-connector-java-8.0.26.jar').getOrCreate()

In [14]:
spark

#### 일별농업기상 데이터

In [83]:
path = './data/일별농업기상/'
file_list = os.listdir(path)
file_list = sorted(file_list)
len(file_list)

462

In [102]:
### raw data 최신화

yesterday = datetime.today() - timedelta(1)
start = end = yesterday.strftime("%Y-%m-%d")

path = './data/일별농업기상/'
file_list = os.listdir(path)
file_list = sorted(file_list)

n_file = []
for file in file_list:
    file = file.rstrip()
    if file.endswith('{}.json'.format(start)):
        n_file.append(file)

# 수집한 최신 데이터 기준 dataframe 열 생성

with open(path+file_list[66], "r", encoding='utf-8') as json_file:
    json_data = json.load(json_file)
    data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
d = json.loads(data)   
if d['response']['body']['items'] == None:
    print('holy')
    
df_base = pd.DataFrame(d['response']['body']['items']['item'], index=[0]) # ValueError: If using all scalar values, you must pass an index 해결
df = pd.DataFrame(columns = df_base.columns)

# 해당 날짜에 해당하는 모든 json파일 불러와 dataframe생성
for i in range(0, len(n_file)):
    with open(path+n_file[i], "r", encoding='utf-8') as json_file:
        json_data = json.load(json_file)
        data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
    d = json.loads(data)
    if d['response']['body']['items'] == None:
        continue
    else:
        df = df.append(d['response']['body']['items']['item'], ignore_index=True)
        #df = df.sort_values(by=['date'], axis=0)

# raw table에 append
df.to_sql(name='weather_raw', con=db_connection, if_exists='append', index=False)

# db에서 raw data 로드해 확인
dataframe_mysql = spark.read.format("jdbc").options(
    url="jdbc:mysql://localhost:3306/farmoney",
    driver = "com.mysql.cj.jdbc.Driver",
    dbtable = "weather_raw",
    user="hoit",
    password="141008").load()

dataframe_mysql.tail(30)

Exception during reset or similar
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/envs/lab08/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 682, in _finalize_fairy
    fairy._reset(pool)
  File "/home/ubuntu/anaconda3/envs/lab08/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 887, in _reset
    pool._dialect.do_rollback(self)
  File "/home/ubuntu/anaconda3/envs/lab08/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 667, in do_rollback
    dbapi_connection.rollback()
  File "/home/ubuntu/anaconda3/envs/lab08/lib/python3.8/site-packages/pymysql/connections.py", line 480, in rollback
    self._read_ok_packet()
  File "/home/ubuntu/anaconda3/envs/lab08/lib/python3.8/site-packages/pymysql/connections.py", line 443, in _read_ok_packet
    pkt = self._read_packet()
  File "/home/ubuntu/anaconda3/envs/lab08/lib/python3.8/site-packages/pymysql/connections.py", line 692, in _read_packet
    packet_header = self._read_bytes(4)
  File "/home/

[Row(no=1, stn_Code='363844A001', stn_Name='청주시 남일면', date='2021-10-04', temp=21.7, max_Temp=29.0, min_Temp=15.1, hum=81.9, widdir=164.6, wind=0.7, rain=0.0, sun_Time=180.0, sun_Qy=10.6, condens_Time=517.0, gr_Temp=None, soil_Temp=21.7, soil_Wt=24.5),
 Row(no=1, stn_Code='232803A001', stn_Name='평창군 여만리', date='2021-10-04', temp=17.9, max_Temp=25.7, min_Temp=12.0, hum=82.6, widdir=192.7, wind=0.7, rain=0.0, sun_Time=74.0, sun_Qy=7.8, condens_Time=0.0, gr_Temp=17.6, soil_Temp=20.0, soil_Wt=19.1),
 Row(no=1, stn_Code='225874A001', stn_Name='횡성군 공근면', date='2021-10-04', temp=19.2, max_Temp=25.3, min_Temp=13.9, hum=87.6, widdir=164.6, wind=0.7, rain=0.0, sun_Time=0.0, sun_Qy=5.1, condens_Time=81.0, gr_Temp=18.6, soil_Temp=20.8, soil_Wt=28.2),
 Row(no=1, stn_Code='059223A001', stn_Name='강진군 군동면', date='2021-10-05', temp=23.1, max_Temp=30.2, min_Temp=17.4, hum=84.9, widdir=203.9, wind=0.6, rain=0.0, sun_Time=480.0, sun_Qy=19.3, condens_Time=0.0, gr_Temp=24.0, soil_Temp=24.9, soil_Wt=36.0),
 R

In [111]:
### 최신 데이터 전처리

# db에서 raw data 로드

#from pyspark.sql import SparkSession
#spark = SparkSession.builder.getOrCreate()
dataframe_mysql = spark.read.format("jdbc").options(
    url="jdbc:mysql://localhost:3306/farmoney",
    driver = "com.mysql.cj.jdbc.Driver",
    dbtable = "weather_raw",
    user="hoit",
    password="141008").load()
raw = dataframe_mysql.select("*").toPandas()

# 일별농업기상 전처리
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
#import plotly.express as px
import matplotlib.pyplot as plt
import matplotlib
from matplotlib import font_manager, rc
sns.set(style='whitegrid')
import missingno as msno
from datetime import datetime, timedelta
%matplotlib inline

rc('font', family='AppleGothic')
    
matplotlib.rcParams['axes.unicode_minus'] = False

def change_columns_name(df):
    df.rename(columns = {'stn_Code' : '관측지점코드',
                            'stn_Name' : '관측지점명',
                            'date' : '관측시각',
                            'temp' : '기온',
                            'max_Temp' : '최고기온',
                            'min_Temp' : '최저기온',
                            'hum' : '습도',
                            'widdir' : '풍향',
                            'wind' : '풍속',
                            'rain' : '강수량',
                            'sun_Time' : '일조시간',
                            'sun_Qy' : '일사량',
                            'condens_Time' : '결로시간',
                            'gr_Temp' : '초상온도',
                            'soil_Temp' : '지중온도',
                            'soil_Wt' : '토양수분보정값'
                            }, inplace = True)
def pick_columns(df):
    #df.drop(['Unnamed: 0','no','관측지점코드'],axis=1,inplace=True)
    df.drop(['no','관측지점코드'],axis=1,inplace=True)
    
df = raw.copy()
change_columns_name(df)
pick_columns(df)
df['관측시각'] = pd.to_datetime(df['관측시각'], format="%Y-%m-%d")
df['연도'] = df['관측시각'].dt.year
df['월'] = df['관측시각'].dt.month
df['일'] = df['관측시각'].dt.day
df_2020 = df[(df['연도'] != 2021)]

df_temp = df.copy()

df_temp = df_temp[(df_temp['관측지점명'] == '진도군 군내면')|(df_temp['관측지점명'] == '신안군 압해읍')|(df_temp['관측지점명'] == '부산시 강서구')|(df_temp['관측지점명'] == '김해시 전하동')
|(df_temp['관측지점명'] == '평창군 여만리')|(df_temp['관측지점명'] == '횡성군 공근면')|(df_temp['관측지점명'] == '남양주 진건읍')|(df_temp['관측지점명'] == '이천시 중리동')]


feature_list = ['기온','최고기온','최저기온','습도','풍향','풍속','강수량','일조시간','일사량','결로시간','초상온도','지중온도','토양수분보정값']

# 0도 이하는 이상치로 판단
for i in [5,6,7,8,9]:
    print(i)
    temp_ds = \
        df_temp.loc[
            df_temp['월'] == i
        ].copy()
    temp_ds['기온'] = temp_ds['기온'].apply(lambda x: np.NaN if (x <= 0) else x)
    temp_index = temp_ds.index
    df_temp.loc[temp_index, "기온"] = temp_ds['기온']

# -25도 이하, 20도 초과는 이상치로 판단
for i in [11,12,1,2]: # 겨울
    temp_ds = \
        df_temp.loc[
            df_temp['월'] == i
        ].copy()
    temp_ds['기온'] = temp_ds['기온'].apply(lambda x: np.NaN if (x <= -25) |(x > 20) else x)
    temp_index = temp_ds.index
    df_temp.loc[temp_index, "기온"] = temp_ds['기온']
    
df_final = df_temp.copy()

def change_new(df, feature_list, loca_list):
    ds = df.copy()
    new_ds = pd.DataFrame()
    na_count = 0 # 결측치 처리 개수
    total_na = 0 # 결측치 처리 전 결측치 개수
    for loca_l in loca_list:
        main_ds = ds[ds["관측지점명"] == loca_l[0]].copy()
        sub_ds = ds[ds["관측지점명"] == loca_l[1]].copy()
        
        # 관측시각이 중복되는 행을 삭제함
        main_before = len(main_ds)
        main_ds.drop_duplicates(['관측시각'], inplace=True)
        main_after = len(main_ds)
        print(f"관측시각이 중복되는 값 {main_after - main_before}개 삭제")
        sub_ds.drop_duplicates(['관측시각'], inplace=True)
        
        for feature in feature_list:
            before_na = len(main_ds.loc[main_ds[feature].isna()])
            temp_dates = \
                main_ds.loc[main_ds[feature].isna(), '관측시각'].astype('str').values.tolist()
            total_na += len(temp_dates)
            for date in temp_dates:
                temp_change = sub_ds.loc[sub_ds['관측시각'] == date, feature].values
                main_ds.loc[main_ds['관측시각'] == date, feature] = temp_change
            after_na = len(main_ds.loc[main_ds[feature].isna()])
            change_na = before_na - after_na
            na_count += change_na
            if change_na == 0:
                print(f"{feature} -> 변경 불가!!!, 잔여na: {after_na}개")
            else:
                print(f"{feature} -> 변경 완료: {change_na}개, 잔여na: {after_na}개")
        new_ds = new_ds.append(main_ds)

    print(f"결측치 처리 전 결측치 개수: {total_na}개")
    print(f"총 결측치 처리 완료 개수: {na_count}개")
    return new_ds

loca_list = [
    ['진도군 군내면', '신안군 압해읍'], 
    ['부산시 강서구', '김해시 전하동'], 
    ['평창군 여만리', '횡성군 공근면'], 
    ['남양주 진건읍', '이천시 중리동']
]
df_final_final = change_new(df_final, feature_list, loca_list)

df_final_final.drop(['초상온도','토양수분보정값', '풍향', '지중온도'],axis=1,inplace=True)

na_f_list = df_final_final.isnull().sum().keys()[(df_final_final.isnull().sum() != 0).values].tolist()

for feature in na_f_list:
    if feature == "기온":
        df_final_final.loc[df_final_final[feature].isna(), feature] = \
            (df_final_final.loc[df_final_final[feature].isna(), "최고기온"] / \
                df_final_final.loc[df_final_final[feature].isna(), "최저기온"])
    elif (feature == "일조시간") | (feature == "일사량") | (feature == "결로시간"):
        df_final_final[feature].fillna(0, inplace=True)
    else:
        print(df_final_final[feature].isna().sum())
        if df_final_final[feature].isna().sum() != 0:
            raise Exception(f"정의되지 않은 Features[{feature}]")
    
#recent = ['2021-10-01','2021-10-02','2021-10-04']
#df_recent = df_final_final[df_final_final['관측시각'].isin(recent)].sort_values(by=['관측시각'], axis=0)
yesterday = datetime.today() - timedelta(1)
start = end = yesterday.strftime("%Y-%m-%d")
df_recent = df_final_final[df_final_final['관측시각'] == start]

# raw table에 append
df_recent.to_sql(name='weather', con=db_connection, if_exists='append', index=False)

# db에서 raw data 로드해 확인
dataframe_mysql = spark.read.format("jdbc").options(
    url="jdbc:mysql://localhost:3306/farmoney",
    driver = "com.mysql.cj.jdbc.Driver",
    dbtable = "weather",
    user="hoit",
    password="141008").load()

dataframe_mysql.tail(30)

5
6
7
8
9
관측시각이 중복되는 값 0개 삭제
기온 -> 변경 완료: 2개, 잔여na: 1개
최고기온 -> 변경 완료: 2개, 잔여na: 0개
최저기온 -> 변경 완료: 2개, 잔여na: 0개
습도 -> 변경 완료: 2개, 잔여na: 0개
풍향 -> 변경 완료: 16개, 잔여na: 0개
풍속 -> 변경 완료: 2개, 잔여na: 0개
강수량 -> 변경 완료: 10개, 잔여na: 0개
일조시간 -> 변경 불가!!!, 잔여na: 2개
일사량 -> 변경 완료: 2개, 잔여na: 0개
결로시간 -> 변경 완료: 2개, 잔여na: 0개
초상온도 -> 변경 불가!!!, 잔여na: 2282개
지중온도 -> 변경 완료: 5개, 잔여na: 0개
토양수분보정값 -> 변경 완료: 28개, 잔여na: 0개
관측시각이 중복되는 값 -10개 삭제
기온 -> 변경 완료: 3개, 잔여na: 0개
최고기온 -> 변경 완료: 3개, 잔여na: 0개
최저기온 -> 변경 완료: 3개, 잔여na: 0개
습도 -> 변경 완료: 70개, 잔여na: 0개
풍향 -> 변경 완료: 200개, 잔여na: 0개
풍속 -> 변경 완료: 3개, 잔여na: 0개
강수량 -> 변경 완료: 3개, 잔여na: 0개
일조시간 -> 변경 완료: 951개, 잔여na: 0개
일사량 -> 변경 완료: 3개, 잔여na: 0개
결로시간 -> 변경 완료: 3개, 잔여na: 0개
초상온도 -> 변경 불가!!!, 잔여na: 2286개
지중온도 -> 변경 완료: 3개, 잔여na: 0개
토양수분보정값 -> 변경 완료: 3개, 잔여na: 1141개
관측시각이 중복되는 값 -1개 삭제
기온 -> 변경 완료: 3개, 잔여na: 0개
최고기온 -> 변경 완료: 3개, 잔여na: 0개
최저기온 -> 변경 완료: 3개, 잔여na: 0개
습도 -> 변경 완료: 3개, 잔여na: 0개
풍향 -> 변경 완료: 8개, 잔여na: 0개
풍속 -> 변경 완료: 3개, 잔여na: 0개
강수량 -> 변경 완료: 3개, 잔여na: 0개
일조시간 -> 변경 완료: 

관측지점명       관측시각    기온  최고기온  최저기온    습도   풍속   강수량   일조시간   일사량  \
25     진도군 군내면 2015-07-01  21.1  25.1  15.9  86.4  2.0   0.5  125.0  21.6   
28     진도군 군내면 2015-07-02  21.0  26.8  13.9  83.6  0.6   0.0  368.0  22.0   
60     진도군 군내면 2015-07-03  22.2  26.6  16.5  79.7  1.1   0.0  303.0  22.8   
86     진도군 군내면 2015-07-04  20.5  27.3  16.4  83.9  0.7   0.0  162.0  20.7   
118    진도군 군내면 2015-07-05  20.6  25.9  16.3  84.7  0.5   0.0  280.0  18.9   
...        ...        ...   ...   ...   ...   ...  ...   ...    ...   ...   
59484  남양주 진건읍 2021-09-30  19.4  26.9  15.7  85.7  0.1   0.0  402.0  14.1   
59501  남양주 진건읍 2021-10-01  19.5  28.5  14.4  81.9  0.2  27.0  530.0  14.7   
59527  남양주 진건읍 2021-10-02  18.8  26.2  14.1  85.3  0.0   0.0  492.0  13.8   
59555  남양주 진건읍 2021-10-04  20.8  25.7  14.9  88.0  0.4   4.0    0.0   2.8   
59582  남양주 진건읍 2021-10-05  23.7  27.1  19.4  79.7  0.9   0.0   35.0   5.1   

        결로시간    연도   월   일  
25     615.0  2015   7   1  
28     456.0  2015   7   2  
60     209.0  2015   7   3  
86     690.0  2015   7   4  
118    734.0  2015   7   5  
...      ...   ...  ..  ..  
59484    0.0  2021   9  30  
59501    0.0  2021  10   1  
59527    0.0  2021  10   2  
59555    0.0  2021  10   4  
59582    0.0  2021  10   5  

[9144 rows x 14 columns]

In [54]:
df_weather_n = spark.read.option("multiline","true").json('./data/일별농업기상/{}'.format(n_file[0]))
df_weather_n = df_weather_n.select('response.body.items.item')
df_weather_n = df_weather_n\
.withColumn('관측지점코드',col('item.stn_Code'))\
.withColumn('관측지점명',col('item.stn_Name'))\
.withColumn('관측시각',col('item.date'))\
.withColumn('기온',col('item.temp'))\
.withColumn('최고기온',col('item.max_temp'))\
.withColumn('최저기온',col('item.min_temp'))\
.withColumn('습도',col('item.hum'))\
.withColumn('풍향',col('item.widdir'))\
.withColumn('풍속',col('item.wind'))\
.withColumn('강수량',col('item.rain'))\
.withColumn('일조시간',col('item.sun_Time'))\
.withColumn('일사량',col('item.sun_Qy'))\
.withColumn('결로시간',col('item.condens_Time'))\
.withColumn('초상온도',col('item.gr_Temp'))\
.withColumn('지중온도',col('item.soil_Temp'))\
.withColumn('토양수분보정값',col('item.soil_Wt'))

df_weather_n = df_weather_n.drop('item','초상온도','토양수분보정값','풍향','지중온도')
df_weather_n.show()

+------------+-------------+----------+----+--------+--------+----+----+------+--------+------+--------+
|관측지점코드|   관측지점명|  관측시각|기온|최고기온|최저기온|습도|풍속|강수량|일조시간|일사량|결로시간|
+------------+-------------+----------+----+--------+--------+----+----+------+--------+------+--------+
|  059223A001|강진군 군동면|2021-10-01|22.5|    29.1|    18.4|  84| 0.6|     0|     521|  20.2|       0|
+------------+-------------+----------+----+--------+--------+----+----+------+--------+------+--------+



#### 시장도매인가격 데이터

In [179]:
### 최신 raw data db에 추가

yesterday = datetime.today() - timedelta(1)
dates = yesterday.strftime("%Y%m%d")

path = './data/일별농업기상/'
file_list = os.listdir(path)
file_list = sorted(file_list)

n_file = []
for file in file_list:
    file = file.rstrip()
    if file.endswith('{}.json'.format(dates)):
        n_file.append(file)

# 시장도매인 가격 pandas dataframe 생성
with open(path+file_list[1000], "r", encoding='utf-8') as json_file:
    json_data = json.load(json_file)
    data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
d = json.loads(data)   
if d['lists']['list_total_count'] == 0:
    print('holy')

df_base = pd.DataFrame(d['lists']['list'])
columns = ['DELNG_DE'] + list(df_base.columns)
#columns = list(df_base.columns)
df = pd.DataFrame(columns = columns)

for i in range(0, len(n_file)):
    with open(path+n_file[i], "r", encoding='utf-8') as json_file:
        json_data = json.load(json_file)
        data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
    d = json.loads(data)
    if d['lists']['list_total_count'] != '4':
        #print(file_list[i][4:12])
        #print(i, d['lists']['list_total_count'])
        continue
    elif d['lists']['list_total_count'] == '4':
        df = df.append(d['lists']['list'], ignore_index=True)
        df['DELNG_DE'].loc[len(df['DELNG_DE'])-4:len(df['DELNG_DE'])] = n_file[i][4:12]

# raw table에 append
df.to_sql(name='market_wholesale', con=db_connection, if_exists='append', index=False)

# db에서 raw data 로드해 확인
dataframe_mysql = spark.read.format("jdbc").options(
    url="jdbc:mysql://localhost:3306/farmoney",
    driver = "com.mysql.cj.jdbc.Driver",
    dbtable = "market_wholesale",
    user="hoit",
    password="141008").load()

dataframe_mysql.tail(10)

[Row(_c0=None, DELNG_DE=20211002, PUM_NM='대파(일반)', G_NAME='보통', U_NAME='1 kg단', MI_P='950', MA_P='1,100', AV_P='1,025', PAV_P='0', FLUC='1,025', D_MARK=0.0, J_7='1,335', J_7_MARK=76.8),
 Row(_c0=None, DELNG_DE=20211002, PUM_NM='대파(일반)', G_NAME='하', U_NAME='1 kg단', MI_P='700', MA_P='950', AV_P='870', PAV_P='0', FLUC='870', D_MARK=0.0, J_7='998', J_7_MARK=87.2),
 Row(_c0=None, DELNG_DE=20211004, PUM_NM='대파(일반)', G_NAME='특', U_NAME='1 kg단', MI_P='1,900', MA_P='2,100', AV_P='1,955', PAV_P='0', FLUC='1,955', D_MARK=0.0, J_7='1,996', J_7_MARK=97.9),
 Row(_c0=None, DELNG_DE=20211004, PUM_NM='대파(일반)', G_NAME='상', U_NAME='1 kg단', MI_P='1,300', MA_P='1,900', AV_P='1,361', PAV_P='0', FLUC='1,361', D_MARK=0.0, J_7='1,612', J_7_MARK=84.4),
 Row(_c0=None, DELNG_DE=20211004, PUM_NM='대파(일반)', G_NAME='보통', U_NAME='1 kg단', MI_P='1,100', MA_P='1,300', AV_P='1,160', PAV_P='0', FLUC='1,160', D_MARK=0.0, J_7='1,258', J_7_MARK=92.2),
 Row(_c0=None, DELNG_DE=20211004, PUM_NM='대파(일반)', G_NAME='하', U_NAME='1 kg

#### 도매시장 경매가격 데이터

In [191]:
path = './data/도매시장경매가격/'
file_list = os.listdir(path)
file_list = sorted(file_list)
len(file_list)

4112

In [ ]:
### raw data 최신화

yesterday = datetime.today() - timedelta(1)
start = end = yesterday.strftime("%Y%m%d")

path = './data/도매시장경매가격/'
file_list = os.listdir(path)
file_list = sorted(file_list)

nn_file = []
for file in file_list:
    file = file.rstrip()
    if file.startswith('{}_'.format(start)):
        n_file.append(file)

# 수집한 최신 데이터 기준 dataframe 열 생성

# 도매시장경매가격 pandas dataframe 생성

with open(path+file_list[62], "r", encoding='utf-8') as json_file:
    json_data = json.load(json_file)
    data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
d = json.loads(data)   
if d['Grid_20180118000000000581_1']['row'] == None:
    print('holy')

df_base = pd.DataFrame(d['Grid_20180118000000000581_1']['row'])
df = pd.DataFrame(columns = df_base.columns)

# 해당 날짜에 해당하는 모든 json파일 불러와 dataframe생성
df = pd.DataFrame(columns = df_base.columns)
for i in range(0, len(file_list)):
    with open(path+file_list[i], "r", encoding='utf-8') as json_file:
        json_data = json.load(json_file)
        data = json.dumps(json_data, indent = '\t', ensure_ascii=False)
    d = json.loads(data)
    if d['Grid_20180118000000000581_1']['row'] == None:
        continue
    else:
        df = df.append(d['Grid_20180118000000000581_1']['row'], ignore_index=True)
        df = df.sort_values(by=['DELNG_DE'], axis=0)

# raw table에 append
df.to_sql(name='wholesale_raw', con=db_connection, if_exists='append',index=False)

In [ ]:
### 최신 데이터 전처리

# db에서 raw data 로드

#from pyspark.sql import SparkSession
#spark = SparkSession.builder.getOrCreate()
dataframe_mysql = spark.read.format("jdbc").options(
    url="jdbc:mysql://localhost:3306/farmoney",
    driver = "com.mysql.cj.jdbc.Driver",
    dbtable = "wholesale_raw",
    user="hoit",
    password="141008").load()
raw_data = dataframe_mysql.select("*").toPandas()

# 도매시장경매데이터 전처리
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

raw_data.drop(columns=["ROW_NUM"], inplace=True)
feature_list = \
    ["DELNG_DE", "PBLMNG_WHSAL_MRKT_NM", "CPR_NM", 
     "SPCIES_NM", "CPR_USE_SPCIES_NM", "PRICE", "DELNGBUNDLE_QY", 
     "STNDRD", "DELNG_QY", "SANJI_NM", "CPR_USE_SANJI_NM"
    ]
select_data = raw_data[feature_list]

# 원본 데이터 개수
total_n = len(raw_data)

feature_list = \
    ["DELNG_DE","PBLMNG_WHSAL_MRKT_NM","CPR_NM", "CPR_USE_SPCIES_NM",
     "PRICE","DELNGBUNDLE_QY","STNDRD","DELNG_QY","SANJI_NM"
    ]

def pprc_default(ds):
    """
    preprocessing
    """

    # 기본 결측치 및 이상치 수정 부분
    data1 = ds.copy()

    # 산지가 결측이면 법인산지로 대치함
    data1.loc[data1["SANJI_NM"].isna(), "SANJI_NM"] = \
        data1.loc[data1["SANJI_NM"].isna(), "CPR_USE_SANJI_NM"]

    # 산지를 찾을 수 없으면 삭제함
    na_n = len(data1.loc[data1["SANJI_NM"].isna()])
    total_n = len(data1)
    print(f"""산지가 없는 값 비율: {format(na_n / total_n * 100, ".2f")}%""")
    print(f"""산지가 없는 값 개수: {format(na_n, ',')}개 삭제""")
    drop_index = data1.loc[data1["SANJI_NM"].isna(), "SANJI_NM"].index
    data1.drop(drop_index, inplace=True)

    # 가격, 거래단량, 거래량이 하나라도 0인 값은 삭제함
    drop_index = data1.loc[
                    (data1["PRICE"] == 0)
                    | (data1["DELNGBUNDLE_QY"] == 0)
                    | (data1["DELNG_QY"] == 0)
                ].index
    print(
        f"가격, 거래단량, 거래량이 하나라도 0인 값 {format(len(drop_index), ',')}개 삭제"
    )
    data1.drop(drop_index, inplace=True)

    # 대파(일반)만 가져옴
    temp = len(data1.loc[data1["SPCIES_NM"] == "대파(일반)"]) / len(data1) * 100
    print("대파(일반)의 비율: {0:.2f}%".format(temp))
    data2 = data1.loc[data1["SPCIES_NM"] == "대파(일반)"].copy()

    # ["대파(일반)", "대파"]만 가져옴
    # 앞 뒤 공백을 제거하고 결측치는 대파로 대체함
    data2.loc[:, "CPR_USE_SPCIES_NM"] = \
        data2.loc[:, "CPR_USE_SPCIES_NM"]\
            .apply(lambda x: x.split()[0] if not pd.isna(x) else "대파")

    data3 = \
        data2.loc[
            (data2["CPR_USE_SPCIES_NM"] == "대파")
            | (data2["CPR_USE_SPCIES_NM"] == "대파(일반)")
        ].copy()

    # 수입이 포함된 값 삭제
    drop_index = data3.loc[data3["SANJI_NM"].str.contains("수입")].index
    data3.drop(drop_index, inplace=True)
    print(f"""산지에 수입이 포함된 값 {format(len(drop_index), ',')}개 삭제""")

    # 필요없는 feature 삭제
    data3.drop(
        columns=["CPR_USE_SANJI_NM", "SPCIES_NM", "CPR_USE_SPCIES_NM"], 
        inplace=True
    )

    return data3

def fat_finger(ds, p=10000, n=3):
    """
    가격이 p원 이상이고
    1의 자리에서 0이 n개 이상 지속되면 0을 하나 삭제함
    """
    
    temp_ds = ds.copy()
    try:
        temp_index = \
            temp_ds.loc[
                (temp_ds["PRICE"] >= p)
                & (temp_ds["PRICE"] % (10 ** n) == 0)
            ].index
        temp_ds.loc[temp_index, "PRICE"] = \
            temp_ds.loc[temp_index, "PRICE"] / 10
    except:
        pass
    
    print(f"fat finger로 {format(len(temp_index), ',')}개 처리함")

    return temp_ds

def pprc_error(ds):
    # 단위가 kg이 아닌 것 중 단위가 g인 것에서 거래단량이 100 미만인 것을
    # kg을 잘못 입력한 것으로 판단하여 kg으로 변환함
    # 즉 거래 단량이 100g 미만이면 이는 kg이라고 판단함
    # 1개 존재함
    data4 = ds.copy()
    g_error_index = data4.loc[
                        (~(data4["STNDRD"].str.contains("kg")))
                        & (data4["STNDRD"].str.contains('g'))
                        & (data4["DELNGBUNDLE_QY"] < 100)
                    ].index
    data4.loc[g_error_index, "STNDRD"] = \
        data4.loc[g_error_index, "STNDRD"].str.replace("g", 'kg')

    # 단위가 kg이 아닌 것 중 단위가 g인 것 kg으로 변환함
    g_index = data4.loc[
                (~(data4["STNDRD"].str.contains("kg")))
                & (data4["STNDRD"].str.contains('g'))
            ].index
    data4.loc[g_index, "DELNGBUNDLE_QY"] = \
        data4.loc[g_index, "DELNGBUNDLE_QY"].apply(lambda x: x / 1000)
    data4.loc[g_index, "STNDRD"] = \
        data4.loc[g_index, "STNDRD"].str.replace("g", 'kg')

    print(
        f"""거래 단량을 잘못 기입한 개수: {
            format(len(g_error_index) + len(g_index), ',')
        }개"""
    )

    # 취소한 것 중 가격과 거래량 모두 음수이면 가격을 양수로 바꿈
    minus_index = data4.loc[
                    (data4["PRICE"] < 0)
                    & (data4["DELNG_QY"] < 0)
                ].index
    data4.loc[minus_index, "PRICE"] = \
        data4.loc[minus_index, "PRICE"].apply(lambda x: -x)

    # 취소한 것 중 가격이 음수이고 거래량이 양수이면 가격을 양수로 바꾸고 거래량은 음수로 바꿈
    p_minus_index = data4.loc[
                        (data4["PRICE"] < 0)
                        & (data4["DELNG_QY"] > 0)
                    ].index
    data4.loc[p_minus_index, "PRICE"] = \
        data4.loc[p_minus_index, "PRICE"].apply(lambda x: -x)
    data4.loc[p_minus_index, "DELNG_QY"] = \
        data4.loc[p_minus_index, "DELNG_QY"].apply(lambda x: -x)

    print(
        f"""거래가격과 거래량이 음수인 개수: {
            format(len(minus_index) + len(p_minus_index), ',')
        }개"""
    )

    # 거래 단량이 1kg 미만인 데이터의 가격을 1kg 단위로 설정함
    temp_index = \
        data4.loc[
            (data4["DELNGBUNDLE_QY"] < 1)
        ].index
    data4.loc[temp_index, "PRICE"] = \
        (
            data4.loc[temp_index, "PRICE"] / \
                data4.loc[temp_index, "DELNGBUNDLE_QY"]
        )

    # 가격이 10,000원 이상이고, 거래단량이 1이면 fat finger로 판단하여 0을 하나 삭제함
    # 0이 2개 이상일 때만 fat finger로 판단함
    # 258개
    data5 = \
        data4.loc[
            (data4["PRICE"] >= 10000)
            & (data4["DELNGBUNDLE_QY"] == 1)
        ].copy()

    print(f"조건 전체: {format(len(data5), ',')}개 중", end=" ")

    data5 = fat_finger(data5, n=2)
    temp_index = data5.index
    data4.loc[temp_index, "PRICE"] = data5["PRICE"]

    # 0이 1개 미만이면 삭제함
    drop_index = data5.loc[data5["PRICE"] >= 10000].index
    data4.drop(drop_index, inplace=True)
    print(f"거래단량이 1인 것 중 이상치{format(len(drop_index), ',')}개 삭제함")

    # # 거래단량이 0.5이하인 것은 삭제함
    # drop_index = \
    #     data4.loc[
    #         (data4["DELNGBUNDLE_QY"] < 0.5)
    #     ].index
    # data4.drop(drop_index, inplace=True)
    # print(f"{format(len(drop_index), ',')}개 삭제함")

    # 거래단량과 거래량이 이상치 값 처리
    temp_index = \
        data4.loc[
            (data4["DELNGBUNDLE_QY"] > 100)
            & (data4["DELNG_QY"] > 100)
            & (data4["DELNG_QY"] == data4["DELNGBUNDLE_QY"])
        ].index
    data4.loc[temp_index, "DELNGBUNDLE_QY"] = 1

    temp_index = \
        data4.loc[
            (data4["DELNGBUNDLE_QY"] > 1000)
            & (data4["DELNG_QY"] > 100)
        ].index
    data4.loc[temp_index, "DELNGBUNDLE_QY"] /= 1000

    return data4

def base_price(ds, min_q=0.3, max_q=0.7):
    """
    같은 날 거래된 거래단량이 1인 가격 중 
    
    min_q(default=30%)와 max_q(default=70%) 및 
    min_q와 max_q의 평균 가격을 반환함
    """
    data = ds.copy()
    def min_quant(x):
        return x.quantile(min_q)
    def max_quant(x):
        return x.quantile(max_q)

    # 원본 데이터의 날짜 집합
    total_d_unique = data["DELNG_DE"].unique()

    data1 = \
        data.loc[data["DELNGBUNDLE_QY"] == 1, ["DELNG_DE", "PRICE"]].copy()

    # 거래단량이 1이 존재하는 날짜의 집합
    q1_d_unique = data1["DELNG_DE"].unique()
    # 거래단량이 1이 존재하지 않는 날짜의 집합
    null_dates = np.setdiff1d(total_d_unique, q1_d_unique)

    data2 = \
        data1.groupby(
            ["DELNG_DE"], 
        ).agg({"PRICE": [
                ("MIN_Q", min_quant), 
                ("MAX_Q", max_quant),
                ("MAX", "max"),
                ("MIN", "min"),
                ]
                }
            ).copy()
    # DELNG_DE의 col_level 설정
    data2.reset_index(col_level=1, inplace=True)
    # col_level 0 삭제
    data3 = data2.droplevel(level=0, axis=1)
    data3["BASE_P"] = (data3["MIN_Q"] + data3["MAX_Q"]) / 2
    data4 = \
        data3.reindex(
            columns=["DELNG_DE", "MIN", "MIN_Q", "BASE_P", "MAX_Q", "MAX"]
        )

    # 거래단량이 1인 값이 없는 경우 전일과 후일의 평균으로 대체함.
    # 후일이 없는 경우 전일의 값으로 사용함
    data5 = data4.copy()
    for date in null_dates:
        i = 1
        while True:
            if len(data5[data5["DELNG_DE"] == date - i]) != 0:
                before_date = date - i
                break
            else:
                i += 1
        i = 1
        while True:
            if len(data5[data5["DELNG_DE"] == date + i]) != 0:
                after_date = date + i
                break
            else:
                i += 1
                if i == 10:
                    after_date = before_date
                    break
        # 전일과 후일 데이터를 가져와 평균치를 반환함
        before_d = data5[data5["DELNG_DE"] == before_date]
        after_d = data5[data5["DELNG_DE"] == after_date]
        temp = pd.concat([after_d, before_d])
        temp["DELNG_DE"] = date
        temp = temp.groupby(["DELNG_DE"], as_index=False).mean()
        data5 = pd.concat([data5, temp]).copy()
        # print(f"{date}의 기준 가격을 {before_date}와 {after_date}의 평균으로 대체함")

    # 날짜 순으로 정렬한 후 인덱스를 초기화함
    data5.sort_values("DELNG_DE", inplace=True)
    data5.reset_index(drop=True, inplace=True)

    return data5

def change_unit_p(ds, unit=10, min_q=0.3, max_q=0.7):
    """
    거래단량이 unit(default=10)개 미만일 때 
    단위 가격이 기준 가격보다 높을 때 
    거래단량이 2.5 이상인 경우: 단위 가격이 기준 가격의 2배 이상이면 단위가격으로 변경한다
    거래단량이 1 초과 2.5 미만인 경우: 단위 가격이 기준 가격의 1.3배 이상이면 단위가격으로 변경한다
    """
    """
    unit: 정의된 거래단량 이하일 때만 변경함
    min_q: minimum quantile
    max_q: maximum quantile
    """
    data = ds.copy()
    
    price_df = base_price(data, min_q=min_q, max_q=max_q)
    
    # 기준가격을 데이터프레임에 합침
    data = pd.merge(data, price_df, on="DELNG_DE", how="outer")
    # 단위가격 feature 생성
    data["UNIT_P"] = data["PRICE"] / data["DELNGBUNDLE_QY"]
    # 결측치 확인
    temp_len = len(data.loc[data["BASE_P"].isna(), "DELNG_DE"].unique())
    print(f"1개 단위로 거래된 적이 없는 날의 개수: {temp_len}일")
    temp_len = len(data.loc[data["BASE_P"].isna()])
    print(f"1개 단위로 거래된 적이 없는 날의 데이터의 개수: {temp_len}개")

    # 거래단량이 2.5 미만인 경우: 단위 가격이 기준 가격의 1.3배 이상이면 단위가격으로 변경한다
    # 단 변경될 단위가격은 min_q(default=30) 이상이고 max 이하여야 한다.
    temp_index = \
        data.loc[
        (data["DELNGBUNDLE_QY"] > 1)
        & (data["DELNGBUNDLE_QY"] < 2.5)
        & (data["BASE_P"].notnull())
        & (data["PRICE"] > \
            (data["BASE_P"] * data["DELNGBUNDLE_QY"] * 1.3)
          )
        & (data["UNIT_P"] >= data["MIN_Q"])
        & (data["UNIT_P"] <= data["MAX"])
        ].index
    data.loc[temp_index, "PRICE"] = \
        round(data.loc[temp_index, "PRICE"] / \
                data.loc[temp_index, "DELNGBUNDLE_QY"],
              0
        )
    print(f"거래단량이 2.5 미만인 경우 {format(len(temp_index), ',')}개 변경됨")

    # 거래단량이 2.5 이상인 경우: 단위 가격이 기준 가격의 2배 이상이면 단위가격으로 변경한다
    # 단 변경될 단위가격은 min_q(default=30) 이상이고 max 이하여야 한다.
    temp_index = \
        data.loc[
            (data["DELNGBUNDLE_QY"] >= 2.5)
            & (data["DELNGBUNDLE_QY"] < unit)
            & (data["BASE_P"].notnull())
            & (data["PRICE"] > (data["BASE_P"] * 2))
            & (data["UNIT_P"] >= data["MIN_Q"])
            & (data["UNIT_P"] <= data["MAX"])
        ].index
    data.loc[temp_index, "PRICE"] = \
        round(data.loc[temp_index, "PRICE"] / \
              data.loc[temp_index, "DELNGBUNDLE_QY"],
              0
        )
    print(f"거래단량이 2.5 이상인 경우 {format(len(temp_index), ',')}개 변경됨")
    
    # 기준 가격 삭제
    data.drop(
        columns=["BASE_P", "MIN_Q", "MAX_Q", "MAX", "MIN", "UNIT_P"], 
        inplace=True,
    )

    return data

def check_p_range(ds, min_q=0.2, max_q=0.8, price=None):
    """
    price: "unit" => 단위가격 사용

    같은 날 거래된 가격 중 price가

    min_q(default=20%)와 max_q(default=80%) 사이에 있으면
    
    TRUE를 반환한다
    """
    data = ds.copy()

    # 같은 날 거래된 가격 중 MIN_Q와 MAX_Q를 구한다.
    price_df = base_price(data, min_q=min_q, max_q=max_q)
    price_df.drop(columns=["MIN", "BASE_P", "MAX"], inplace=True)
    data = pd.merge(data, price_df, on="DELNG_DE", how="outer")
    # 가격이 MIN_Q와 MAX_Q 사이에 있으면 TRUE를 반환한다.
    if price == "unit":
        data["UNIT_P"] = data["PRICE"] / data["DELNGBUNDLE_QY"]
        p = "UNIT_P"
    else:
        p = "PRICE"
    
    data["Q_RANGE"] = (
        (data["MIN_Q"] <= data[p])
        & (data[p] <= data["MAX_Q"])
    )
    if price == "unit":
        data.drop(columns=["MIN_Q", "MAX_Q", "UNIT_P"], inplace=True)
    else:
        data.drop(columns=["MIN_Q", "MAX_Q"], inplace=True)
    
    return data


def pprc_price(ds, price=9000, weight=1.3):
    """
    가격 이상치를 처리함

    price: 를 초과하는 가격일 때 이상치로 판단함

    weight: 거래단량이 1인 것의 MAX 값에 곱할 가중치
    """
    data = ds.copy()

    # 단위가격이 quantile 범위 내에 드는지 확인하는 feature 'Q_RANGE'생성
    data = check_p_range(data, min_q=0.2, max_q=0.8, price="unit")

    # 가격이 9,000원 이상이고 Q_RANGE가 TRUE이면 단위가격으로 변경함
    temp_index = \
        data.loc[
            (data["Q_RANGE"])
            & (data["PRICE"] > price)
        ].index
    data.loc[temp_index, "PRICE"] = \
        data.loc[temp_index, "PRICE"] / data.loc[temp_index, "DELNGBUNDLE_QY"]
    print(f"단위가격으로 {format(len(temp_index), ',')}개 처리함")

    # 같은 날 거래된 가격 중 거래단량이 1인 것의 MAX를 구한다.
    price_df = base_price(data)
    price_df.drop(columns=["MIN_Q", "BASE_P", "MAX_Q"], inplace=True)
    data = pd.merge(data, price_df, on="DELNG_DE", how="outer")

    # 가격이 9,000원 이상이고 Q_RANGE가 FALSE이며,
    # 단위가격이 MAX * weight(default=1.3)을 초과할 때 
    # 0이 2개 연속이면 fat_finger로 처리함
    data["UNIT_P"] = data["PRICE"] / data["DELNGBUNDLE_QY"]
    temp_ds = \
        data.loc[
            ~(data["Q_RANGE"])
            & (data["PRICE"] > price)
            & (data["UNIT_P"] > data["MAX"] * weight)
        ]
    temp_index = temp_ds.index
    temp_ds = fat_finger(temp_ds, p=price, n=2)
    data.loc[temp_index] = temp_ds

    # 가격이 9,000원 이상이고 Q_RANGE가 FALSE일 때 
    # 단위가격이 MIN보다 크거나 같고 MAX * weight + 0.5(default=1.8) 이내이면
    # 단위가격으로 변경함
    temp_index = \
        data.loc[
            ~(data["Q_RANGE"])
            & (data["PRICE"] > price)
            & (data["UNIT_P"] >= data["MIN"])
            & (data["UNIT_P"] <= data["MAX"] * (weight + 0.5))
        ].index
    data.loc[temp_index, "PRICE"] = data.loc[temp_index, "UNIT_P"]
    print(f"단위가격으로 {format(len(temp_index), ',')}개 처리함")

    data = check_p_range(data, min_q=0.2, max_q=0.8, price="unit")
    # 가격이 9,000원 이상이고 Q_RANGE가 FALSE이며, 
    # 0이 2개 연속이면 fat_finger로 처리함
    data["UNIT_P"] = data["PRICE"] / data["DELNGBUNDLE_QY"]
    temp_ds = \
        data.loc[
            ~(data["Q_RANGE"])
            & (data["PRICE"] > price)
        ]
    temp_index = temp_ds.index
    temp_ds = fat_finger(temp_ds, p=price, n=2)
    data.loc[temp_index] = temp_ds

    drop_index = \
        data.loc[
            ~(data["Q_RANGE"])
            & (data["PRICE"] > price)
        ].index
    data.drop(drop_index, inplace=True)
    print(f"{format(len(drop_index), ',')}개 삭제함")

    data.drop(columns=["Q_RANGE", "MIN", "MAX", "UNIT_P"], inplace=True)

    return data

# 창원내서도매시장 이상치로 인해 drop
drop_index = select_data[
    select_data["PBLMNG_WHSAL_MRKT_NM"] == "창원내서도매시장"
].index
data0 = select_data.drop(drop_index).copy()

data1 = pprc_default(data0)
data2 = pprc_error(data1)
data3 = change_unit_p(data2, unit=float("inf"))
data4 = change_unit_p(data3, unit=float("inf"), min_q=0.01, max_q=0.01)
data5 = pprc_price(data4, price=9000, weight=1.3)

data5["VOLUME"] = data5["DELNG_QY"] * data5["DELNGBUNDLE_QY"]


def loca_change(ds):
    """
    시군구를 정의된 양식에 따라 변경함
    """

    data = ds.copy()

    global error_list
    error_list = set()
    sido_dict = {
        "경기도": "경기",
        "강원도": "강원",
        "경상북도": "경북",
        "경상남도": "경남",
        "충청북도": "충북",
        "충청남도": "충남",
        "전라북도": "전북",
        "전라남도": "전남",
        "서울특별시": "서울",
        "세종특별시": "세종",
        "부산광역시": "부산",
        "울산광역시": "울산",
        "대구광역시": "대구",
        "대전광역시": "대전",
        "제주특별자치도": "제주",
        "제주도": "제주",
        "광주광역시": "광주",
        "인천광역시": "인천",
    }

    def check_si_gun_gu(text):
        """
        시/군/구가 포함되어 있는지 확인한다.
        """
        if (text[-1] == "시") | (text[-1] == "군") | (text[-1] == "구"):
            return True
        else:
            return False

    def check_double(text):
        """
        광역자치단체 명이 중복되어 있는지 확인한다.
        """
        temp_list = sido_dict.values()
        if text in temp_list:
            return True
        else:
            return False

    def loca_name_change(data):
        """
        광역자치단체 명과 시군구 명을 통일한다.
        """

        temp = list(filter(None, data.split(" ")))
        
        # 광역시도명 짧게 변경
        if temp[0] in sido_dict.keys():
            temp[0] = sido_dict[temp[0]]
        
        # 시군구 짧게 변경
        if len(temp) >= 2:
            if check_double(temp[1]):
                temp[1] = temp[2]

            if check_si_gun_gu(temp[1]): 
                res = temp[0] + " " + temp[1][:-1] # 시군구 삭제하고 결합
            else:
                res = temp[0] + " " + temp[1]
        else:
            res = temp[0]

        return res

    def merge_loca(data):
        """
        사전에 정의한 지리적으로 근접한 지역을 합침.
        """
        def exception_process(text):
            """
            예외적인 지역 이름을 처리함.
            """
            # 리스트로 들어오면 문자열로 변경함
            text = ' '.join(text)
            exception_dict = {"완주": "전라남도(진도+)",
                            }

            # 시군구가 포함되어 있으면 시군구로 변경함
            for area in area_dict.keys():
                if area in text:
                    text = area_dict[area]
                    # 종료
                    return text

            # 예외 처리 항목에 포함되어 있으면 해당 값으로 변경함
            for area in exception_dict.keys():
                if area in text:
                    text = exception_dict[area]
                    return text
            
            # 광역지자체가 포함되어 있으면 광역지자체로 변경함
            for area in metro_area_dict.keys():
                if area in text:
                    text = metro_area_dict[area]
                    return text
            
            # # 처리되지 않았으면 메세지 출력함 
            # print('='*79)
            # print(f"{text} 처리 필요함!!!".center(60, ' '))
            # print('='*79)

            # 처리되지 않은 지역 리스트 추가
            global error_list
            error_list.update([text])
            
            return text

        area_dict = {
            # "신안": "전라남도(진도+)",
            # "진도": "전라남도(진도+)",
            # "해남": "전라남도(진도+)",
            # "안양": "경기도(구리+)",
            # "구리": "경기도(구리+)",
            # "남양주": "경기도(구리+)",
            # "하남": "경기도(구리+)",
            # "성남": "경기도(구리+)",
            # "광주": "경기도(구리+)",
            # "수원": "경기도(구리+)",
            # "안성": "경기도(구리+)",
            # "오산": "경기도(구리+)",
            # "평택": "경기도(구리+)",
            # "이천": "경기도(구리+)",
            # "여주": "경기도(구리+)",
            # "양평": "경기도(구리+)",
            # "용인": "경기도(구리+)",
            # "화성": "경기도(구리+)",
            # "고양": "경기도(구리+)",
            # "파주": "경기도(구리+)",
            # "양주": "경기도(구리+)",
            # "동두천": "경기도(구리+)",
            # "연천": "경기도(구리+)",
            # "포천": "경기도(구리+)",
            # "가평": "경기도(구리+)",
            # "의정부": "경기도(구리+)",
            # "김포": "경기도(구리+)",
            # "부천": "경기도(구리+)",
            # "안산": "경기도(구리+)",
            # "군포": "경기도(구리+)",
            # "광명": "경기도(구리+)",
            # "시흥": "경기도(구리+)",
            # "평창": "강원도(평창+)",
        }
        metro_area_dict = {
            "경남": "경상남도(부산+)",
            "부산": "경상남도(부산+)",
            "울산": "경상남도(부산+)",
            "전남": "전라남도(진도+)",
            "광주": "전라남도(진도+)",
            "경기": "경기도(구리+)",
            "강원": "강원도(평창+)",
            # "전북": "전라북도",
            # "충북": "충청북도 청주+",
            # "충남": "충청남도 아산+",
            # "경북": "경상북도 구미+(+대구)",
            # "대전": "충청북도 청주+",
            # "세종": "충청북도 청주+",
            # "대구": "경상북도 구미+(+대구)",
            # "서울": "경기도(구리+)",
            # "인천": "경기도(구리+)",
        }

        temp = list(filter(None, data.split(" ")))
        
        if len(temp) >= 2:
            if temp[1] in area_dict.keys():
                res = area_dict[temp[1]]
            elif temp[0] in metro_area_dict.keys():
                res = metro_area_dict[temp[0]]
            else:
                res = exception_process(temp)
        else:
            if temp[0] in metro_area_dict.keys():
                res = metro_area_dict[temp[0]]
            else:
                res = exception_process(temp)

        return res

    data.loc[:, "SANJI_NM"] = data.loc[:, "SANJI_NM"].apply(loca_name_change)

    # 생산량이 미미한 지역 제거함
    drop_loca_list = \
        ["제주", "전북", "충북", "충남", "경북", "대전", "세종", "대구", "서울", "인천"]
    for d_loca in drop_loca_list:
        drop_index = data[data["SANJI_NM"].str.contains(d_loca)].index
        data.drop(drop_index, inplace=True)

    # 산지 명을 정의한대로 설정
    data.loc[:, "SANJI_NM"] = data.loc[:, "SANJI_NM"].apply(merge_loca)

    # 정의되지 않은 지역 리스트 출력
    print(f"처리가 필요한 지역 리스트: {error_list}")
    for er_loca in error_list:
        drop_index = data[data["SANJI_NM"].str.contains(er_loca)].index
        data.drop(drop_index, inplace=True)

    # 정의된 지역의 개수 = 13개
    # data["SANJI_NM"].unique()
    # data["SANJI_NM"].value_counts()
    print(f"""데이터 내 지역의 개수: {data["SANJI_NM"].nunique()}개""")

    return data

data6 = loca_change(data5)

# 분석용 table db업로드
data5.to_sql(name='wholesale1', con=db_connection, if_exists='append', index=False)

# 모델링용 table db업로드
data6.to_sql(name='wholesale2', con=db_connection, if_exists='append', index=False)

# db에서 로드해 확인
dataframe_mysql = spark.read.format("jdbc").options(
    url="jdbc:mysql://localhost:3306/farmoney",
    driver = "com.mysql.cj.jdbc.Driver",
    dbtable = "wholesale1",
    user="hoit",
    password="141008").load()

dataframe_mysql.tail(30)

In [ ]:
# 최종적으로 학습된 모델에 넣을 데이터 생성

############################################################
FUTURE_SIZE = 1

# 1회 학습에 포함될 과거 N 단위 기간
WINDOW_SIZE = 20 # 약 5개월

# 테스트 데이터에 포함될 단위 기간 수
TEST_SIZE = 30 
TARGET = "VOLUME"
############################################################

############################################################
LOCATION = ["gangwon", "gyengnam", "gyeonggi", "jeonnam"]
############################################################

# 모델용 전처리 후 도매가격데이터
dataframe_mysql = spark.read.format("jdbc").options(
    url="jdbc:mysql://localhost:3306/farmoney",
    driver = "com.mysql.cj.jdbc.Driver",
    dbtable = "wholesale2",
    user="hoit",
    password="141008").load()
wholesale_raw = dataframe_mysql.select("*").toPandas()

# 모델용 전처리 후 날씨 데이터
dataframe_mysql = spark.read.format("jdbc").options(
    url="jdbc:mysql://localhost:3306/farmoney",
    driver = "com.mysql.cj.jdbc.Driver",
    dbtable = "weather",
    user="hoit",
    password="141008").load()
weather_raw = dataframe_mysql.select("*").toPandas()
try:
    weather_raw.drop(columns=["_c0"], inplace=True)
except:
    pass

def pr_weather(ds):
    """
    기상 데이터 전처리

    생육 적정 온도에서 벗어나는 값, 
    발아 한계 기온에서 벗어나는 값, 
    태풍 여부 등을 처리함
    """
    weather = ds.copy()
    weather["관측시각"] = pd.to_datetime(weather["관측시각"], format="%Y-%m-%d")
    weather.drop(columns=["연도", "월", "일"], inplace=True)
    weather.sort_values("관측시각", inplace=True)
    
    # 생육 적정온도 15 ~ 25도
    # 최적기온을 벗어나는 값을 최저기온, 평균기온, 최고기온에 따라
    # 가중치 값을 달리함
    weather["!최적기온"] = 0 
    temp_index = \
        weather[
            (weather["최저기온"] < 15)
            | (weather["최고기온"] > 25)
        ].index
    weather.loc[temp_index, "!최적기온"] += 1

    temp_index = \
        weather[
            (weather["기온"] < 15)
            | (weather["기온"] > 25)
        ].index
    weather.loc[temp_index, "!최적기온"] += 5

    temp_index = \
        weather[
            (weather["최고기온"] < 15)
            | (weather["최저기온"] > 25)
        ].index
    weather.loc[temp_index, "!최적기온"] += 10

    # 발아 한계 기온
    # 발아 한계 기온에서 벗어나는 값을 최저기온, 평균기온, 최고기온에 따라
    # 가중치 값을 달리함
    weather["불량기온"] = 0 
    temp_index = \
        weather[
            (weather["최저기온"] < 0)
            | (weather["최고기온"] > 33)
        ].index
    weather.loc[temp_index, "불량기온"] += 1

    temp_index = \
        weather[
            (weather["기온"] < 0)
            | (weather["기온"] > 33)
        ].index
    weather.loc[temp_index, "불량기온"] += 5

    temp_index = \
        weather[
            (weather["최고기온"] < 0)
            | (weather["최저기온"] > 33)
        ].index
    weather.loc[temp_index, "불량기온"] += 10

    # 강수량이 90을 초과하고 풍속이 4 이상이면 태풍이라고 판단함
    weather["태풍"] = 0
    weather.loc[
        (weather["강수량"] > 90) & (weather["풍속"] > 4),
        "태풍"
    ] = 1    

    agg_dict = {
        "기온": [("기온", "mean")],
        "일조시간": [("일조시간", "sum")],
        "일사량": [("일사량", "sum")],
        "결로시간": [("결로시간", "sum")],
        "!최적기온": [("!최적기온", "sum")],
        "불량기온": [("불량기온", "sum")],
        "태풍": [("태풍", "sum")],
    }
    
    # 연도와 주번호 생성
    weather["WEEK_N"] = weather["관측시각"].dt.strftime('%U')
    weather["YEAR"] = weather["관측시각"].dt.year
    weather_g = \
        weather.groupby(["관측지점명", "YEAR", "WEEK_N"]).agg(agg_dict)
    weather_g.reset_index(col_level=1, inplace=True)
    weather_g = weather_g.droplevel(level=0, axis=1)
    
    # 지역명 변경
    loca_dict = {
        "진도군 군내면": "전라남도(진도+)",
        "부산시 강서구": "경상남도(부산+)",
        "평창군 여만리": "강원도(평창+)",
        "남양주 진건읍": "경기도(구리+)"
    }

    for loc in loca_dict.keys():
        weather_g.loc[weather_g["관측지점명"] == loc, "관측지점명"] = loca_dict[loc]

    return weather_g

def pr_main(ds, drop_leap_year=True):
    """
    기본적인 데이터 처리

    drop_leap_year=True: 윤년(leap year) 2월 29일 데이터 삭제
    """
    main = ds.copy()
    
    main["DELNG_DE"] = pd.to_datetime(main["DELNG_DE"], format="%Y%m%d")
    drop_columns = \
        [
         "PBLMNG_WHSAL_MRKT_NM", "CPR_NM", 
         "DELNGBUNDLE_QY", "STNDRD", "DELNG_QY"
        ]
    main.drop(columns=drop_columns, inplace=True)
    main.reset_index(drop=True, inplace=True)

    # 윤년(leap year) 2월 29일 삭제
    if drop_leap_year:
        drop_index = \
            main.loc[
                (main["DELNG_DE"].dt.month == 2)
                & (main["DELNG_DE"].dt.day == 29)
            ].index
        main.drop(drop_index, inplace=True)
    else:
        pass
    
    return main

# 메인 데이터 주별 그룹화
def main_to_week(ds):
    main = ds.copy()
    
    main.sort_values("DELNG_DE", inplace=True)
    # 주번호(일요일 기준)
    main["WEEK_N"] = main["DELNG_DE"].dt.strftime('%U')
    # 연도
    main["YEAR"] = main["DELNG_DE"].dt.year
    
    # 월 더미 추가
    month_dumm = pd.get_dummies(main["DELNG_DE"].dt.month_name())
    main = pd.concat([main, month_dumm], axis=1)

    # 그룹화
    def q80(x):
        return x.quantile(0.8)
    def q20(x):
        return x.quantile(0.2)

    agg_dict = {
        "PRICE": [
            # ("P_MIN", "min"),
            # ("P_MAX", "max"),
            ("P_MEDIAN", np.median),
            ("P_MEAN", np.mean),
            ("P_STD", np.std),
            ("P_Q80", q80),
            ("P_Q20", q20),
        ],
        "VOLUME": [
            ("VOLUME", "sum"),
            ("V_STD", np.std)
        ],
        "January": [("January", "mean")],
        "February": [("February", "mean")],
        "March": [("March", "mean")],
        "April": [("April", "mean")],
        "May": [("May", "mean")],
        "June": [("June", "mean")],
        "July": [("July", "mean")],
        "August": [("August", "mean")],
        "September": [("September", "mean")],
        "October": [("October", "mean")],
        "November": [("November", "mean")],
        "December": [("December", "mean")]
    }
    main_g = main.groupby(["SANJI_NM", "YEAR", "WEEK_N"]).agg(agg_dict).copy()
    main_g.reset_index(col_level=1, inplace=True)
    main_g = main_g.droplevel(level=0, axis=1)

    # 표준편차 결측값 0으로 대치
    main_g["P_STD"].fillna(0, inplace=True)
    
    # 가격 중앙값 이동평균 
    main_g["P_EMA5"] = main_g["P_MEDIAN"].ewm(5, min_periods=1).mean()
    main_g["P_EMA15"] = main_g["P_MEDIAN"].ewm(15, min_periods=1).mean()
    main_g["P_EMA30"] = main_g["P_MEDIAN"].ewm(30, min_periods=1).mean()
    
    # 거래량 이동평균
    main_g["V_EMA5"] = main_g["VOLUME"].ewm(5, min_periods=1).mean()
    main_g["V_EMA15"] = main_g["VOLUME"].ewm(15, min_periods=1).mean()
    main_g["V_EMA30"] = main_g["VOLUME"].ewm(30, min_periods=1).mean()
    
    return main_g

wholesale = pr_main(wholesale_raw, drop_leap_year=True)
weather = pr_weather(weather_raw)

# 도매 데이터 주단위로 변경
wholesale_weeked = main_to_week(wholesale) 

# 날씨와 도매 데이터 병합
final_df = \
    pd.merge(
        wholesale_weeked, weather, 
        left_on=["YEAR", "WEEK_N", "SANJI_NM"], 
        right_on=["YEAR", "WEEK_N", "관측지점명"],
        how="left"
    )
final_df.drop(columns=["관측지점명", "YEAR", "WEEK_N"], inplace=True)

# # 모델에 넣어 transfrom할 데이터이므로 TARGET은 삭제함
# final_df.drop(columns=["VOLUME"], inplace=True)

# 컬럼 순서 학습 데이터셋과 맞추기
columns_order = ['SANJI_NM', 'P_MEDIAN', 'P_STD', 'P_Q80', 'P_Q20', 'V_STD', 'January', 'February',
       'March', 'April', 'May', 'June', 'July', 'August', 'September',
       'October', 'November', 'December', 'P_EMA5', 'P_EMA15', 'P_EMA30',
       'V_EMA5', 'V_EMA15', 'V_EMA30', '기온', '일조시간', '일사량', '결로시간', '!최적기온',
       '불량기온', '태풍', 'P_MEAN', 'VOLUME']
f_df_ordered = final_df[columns_order].copy()

# 각 산지별로 데이터 분리
gangwon = f_df_ordered[f_df_ordered["SANJI_NM"] == "강원도(평창+)"].copy()
gangwon.drop(columns="SANJI_NM", inplace=True)
jeonnam = f_df_ordered[f_df_ordered["SANJI_NM"] == "전라남도(진도+)"].copy()
jeonnam.drop(columns="SANJI_NM", inplace=True)
gyengnam = f_df_ordered[f_df_ordered["SANJI_NM"] == "경상남도(부산+)"].copy()
gyengnam.drop(columns="SANJI_NM", inplace=True)
gyeonggi = f_df_ordered[f_df_ordered["SANJI_NM"] == "경기도(구리+)"].copy()
gyeonggi.drop(columns="SANJI_NM", inplace=True)

# data 저장
data_dict = {
    "gangwon": gangwon, 
    "gyengnam": gyengnam, 
    "gyeonggi": gyeonggi, 
    "jeonnam": jeonnam
}

# 모델에 학습시킨 WINDOW_SIZE만큼의 데이터만 가져오기
for loc in LOCATION:
    globals()["input_model_ds_" + loc] = data_dict[loc][-WINDOW_SIZE:]

# db에 table로 append
input_model_ds_gangwon.to_sql(name='gangwon_pred', con=db_connection, if_exists='append', index=False)
input_model_ds_gyengnam.to_sql(name='gyengnam_pred', con=db_connection, if_exists='append', index=False)
input_model_ds_gyeonggi.to_sql(name='gyeonggi_pred', con=db_connection, if_exists='append', index=False)
input_model_ds_jeonnam.to_sql(name='jeonnam_pred', con=db_connection, if_exists='append', index=False)

In [33]:
# struct
df2 = spark.read.option("multiline","true").json('./data/도매시장경매가격/20210911_1_1000.json')
df2 = df2.select("Grid_20180118000000000581_1.row")
df2.printSchema()

root
 |-- row: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- AUC_SE_CODE: string (nullable = true)
 |    |    |-- AUC_SE_NM: string (nullable = true)
 |    |    |-- BIDTIME: string (nullable = true)
 |    |    |-- CPR_CD: string (nullable = true)
 |    |    |-- CPR_NM: string (nullable = true)
 |    |    |-- CPR_USE_PRDLST_NM: string (nullable = true)
 |    |    |-- CPR_USE_SANJI_CD: string (nullable = true)
 |    |    |-- CPR_USE_SANJI_NM: string (nullable = true)
 |    |    |-- CPR_USE_SPCIES_NM: string (nullable = true)
 |    |    |-- DELNGBUNDLE_QY: double (nullable = true)
 |    |    |-- DELNG_DE: string (nullable = true)
 |    |    |-- DELNG_QY: long (nullable = true)
 |    |    |-- GRAD: string (nullable = true)
 |    |    |-- GRAD_CD: string (nullable = true)
 |    |    |-- ORGNO: string (nullable = true)
 |    |    |-- PBLMNG_WHSAL_MRKT_CD: string (nullable = true)
 |    |    |-- PBLMNG_WHSAL_MRKT_NM: string (nullable = true)
 |    |   

In [44]:
import pyspark.sql.functions as F

def flatten_df(nested_df):
    flat_cols = [c[0] for c in nested_df.dtypes if c[1][:6] != 'struct']
    nested_cols = [c[0] for c in nested_df.dtypes if c[1][:6] == 'struct']

    flat_df = nested_df.select(flat_cols +
                               [F.col(nc+'.'+c).alias(nc+'_'+c)
                                for nc in nested_cols
                                for c in nested_df.select(nc+'.*').columns])
    return flat_df

In [49]:
flatten_df(df2).printSchema()

root
 |-- row: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- AUC_SE_CODE: string (nullable = true)
 |    |    |-- AUC_SE_NM: string (nullable = true)
 |    |    |-- BIDTIME: string (nullable = true)
 |    |    |-- CPR_CD: string (nullable = true)
 |    |    |-- CPR_NM: string (nullable = true)
 |    |    |-- CPR_USE_PRDLST_NM: string (nullable = true)
 |    |    |-- CPR_USE_SANJI_CD: string (nullable = true)
 |    |    |-- CPR_USE_SANJI_NM: string (nullable = true)
 |    |    |-- CPR_USE_SPCIES_NM: string (nullable = true)
 |    |    |-- DELNGBUNDLE_QY: double (nullable = true)
 |    |    |-- DELNG_DE: string (nullable = true)
 |    |    |-- DELNG_QY: long (nullable = true)
 |    |    |-- GRAD: string (nullable = true)
 |    |    |-- GRAD_CD: string (nullable = true)
 |    |    |-- ORGNO: string (nullable = true)
 |    |    |-- PBLMNG_WHSAL_MRKT_CD: string (nullable = true)
 |    |    |-- PBLMNG_WHSAL_MRKT_NM: string (nullable = true)
 |    |   

In [35]:
df2.createOrReplaceTempView('df2')

"ROW_NUM": 1,
"DELNG_DE": "20210911",
"PBLMNG_WHSAL_MRKT_NM": "서울강서도매",
"PBLMNG_WHSAL_MRKT_CD": "110008",
"CPR_NM": "강서청과",
"CPR_CD": "11000803",
"RISENO": "1",
"BIDTIME": "20210911",
"AUC_SE_CODE": "1",
"AUC_SE_NM": "경매",
"ORGNO": "210912010235303",
"PRDLST_NM": "대파",
"PRDLST_CD": "1202",
"SPCIES_NM": "대파(일반)",
"SPCIES_CD": "120201",
"CPR_USE_PRDLST_NM": "",
"CPR_USE_SPCIES_NM": "",
"PRICE": 1100,
"DELNGBUNDLE_QY": 1,
"STNDRD": "kg PE대 1개",
"STNDRD_CD": "120301",
"GRAD": "없음",
"GRAD_CD": "10",
"SHIPMNT_SE_CODE": "",
"SHIPMNT_SE_NM": "",
"SANJI_CD": "100000",
"SANJI_NM": "서울 서울 중구",
"CPR_USE_SANJI_CD": "",
"CPR_USE_SANJI_NM": "",
"DELNG_QY": 310

In [43]:
spark.sql('SELECT * FROM df2').show()

+----------------------+
|                   row|
+----------------------+
|[{1, 경매, 20210911...|
+----------------------+



In [40]:
spark.sql('SELECT row.ROW_NUM, row.DELNG_DE FROM df2').show()

+--------------------+--------------------+
|             ROW_NUM|            DELNG_DE|
+--------------------+--------------------+
|[1, 2, 3, 4, 5, 6...|[20210911, 202109...|
+--------------------+--------------------+



In [ ]:
input_model_ds_gangwon.to_sql(name='gangwon_pred', con=db_connection, if_exists='append', index=False)
input_model_ds_gyengnam.to_sql(name='gyengnam_pred', con=db_connection, if_exists='append', index=False)
input_model_ds_gyeonggi.to_sql(name='gyeonggi_pred', con=db_connection, if_exists='append', index=False)
input_model_ds_jeonnam.to_sql(name='jeonnam_pred', con=db_connection, if_exists='append', index=False)